### PR TITLE
Cleanup transactional batch execute

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

File: src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs [hunk 1]
```diff
@@ -2,43 +2,58 @@
...
+                        operation.Success(result);
+                        continue;
                     }
 
-                    if (result.StatusCode == HttpStatusCode.BadRequest)
-                    {
-                        throw new Exception("Bad request. Quite likely the partition key did not match");
-                    }
+                    // guaranteed to throw
+                    operation.Conflict(result);
```
Comment (boblangley) [inline src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:56]: We throw on the first operation that failed. Does `TransactionalBatch` do the same, or should we somehow aggregate all of the results and throw that?

Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:56]: I preserved the behaviour we already had in place. Happy to do what you suggest. Another PR?

Comment (boblangley) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:56]: Sure.

**Review by boblangley — APPROVED** [review]

**Review by SeanFeldman — APPROVED** [review]

File: src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs (hunk 1 above)
Comment (SeanFeldman) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:56]: We can aggregate status codes of all the operations, but only the failing operation will have the error message. Every operation that did not execute as a result of the faulted operations will only have its status code set to [`424 - FailedDependency`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.transactionalbatch.executeasync?view=azure-dotnet#Microsoft_Azure_Cosmos_TransactionalBatch_ExecuteAsync_System_Threading_CancellationToken_).

> If an operation within the transactional batch fails during execution, no changes from the batch will be committed and the status of the failing operation is made available in the StatusCode. To get more details about the operation that failed, the response can be enumerated - this returns TransactionalBatchOperationResult instances corresponding to each operation in the transactional batch in the order they were added into the transactional batch. For a result corresponding to an operation within the transactional batch, the StatusCode indicates the status of the operation - if the operation was not executed or it was aborted due to the failure of another operation within the transactional batch, the value of this field will be HTTP 424 (Failed Dependency); for the operation that caused the batch to abort, the value of this field will indicate the cause of failure as a HTTP status code.

Given that, I see no point in aggregation.

File: src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs [hunk 2]
```diff
@@ -2,43 +2,58 @@
...
     {
+        internal static async Task Execute(this TransactionalBatchDecorator transactionalBatch, Operation operation)
+        {
+            operation.Apply(transactionalBatch);
+
+            using (var batchOutcomeResponse = await transactionalBatch.Inner.ExecuteAsync().ConfigureAwait(false))
+            {
+                if (batchOutcomeResponse.Count > 1)
+                {
+                    throw new Exception("The transactional batch can only have one operation.");
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: Do I understand correctly that this is very specific for the outbox dispatch case (`SetAsDispatched`)? I.e. no other operations would be expected to be found other than the outbox operation. If that's the case, the exception could be more informative and elaborate on why we're throwing.

Comment (SeanFeldman) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: Also, is it possible to get into this scenario other than us using the API incorrectly?

Comment (boblangley) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: That is how I understand it. Only we would cause this situation.

Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: From the perspective of this code this exception enforces that this method can only ever have a single operation within a transactional batch, nothing prevents this method from being used elsewhere as long as these requirements are fulfilled. That's why I did not mention directly the outbox part

Comment (SeanFeldman) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: Right, what confused me is the wording.
> The transactional batch can only have one operation.

I would be OK if it's used not just with the Outbox but have the following message:
> The transactional batch was intended to be used with a single operation but contained more than one.

Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB/TransactionalBatchDecoratorExtensions.cs:17]: Raised https://github.com/Particular/NServiceBus.Persistence.CosmosDB/pull/21

**Pull request merged**